src/utils/symbol_utils.ml

```diff
@@ -8,18 +8,76 @@ type symbol =
   }
 
 
-let find_symbol program name =
+type extern_symbol =
+  {
+    tid : tid;
+    address : string;
+    name : string;
+    cconv : string option;
+    args : (Var.t * Exp.t * intent option) list;
+  }
+
+
+let extern_symbols = ref []
+
+
+let get_project_calling_convention (project : Project.t) : string option =
+  Project.get project Bap_abi.name
+
+
+let build_extern_symbols (project : Project.t) (program : program term) : unit =
+  let parsed_symbols = String.Set.to_list (Cconv.parse_dyn_syms project) in
+  let calling_convention = get_project_calling_convention project in
+  let tid_map = Address_translation.generate_tid_map program in
+  extern_symbols := List.append !extern_symbols (Seq.to_list (Seq.filter_map (Term.enum sub_t program) ~f:(fun s ->
+    let sub_name = Sub.name s in
+    let sub_tid = Term.tid s in
+    match (Stdlib.List.mem sub_name parsed_symbols) with
+    | true -> begin
+        let addr = Address_translation.translate_tid_to_assembler_address_string sub_tid tid_map in
+        let args = Seq.to_list (Seq.map (Term.enum arg_t s) ~f:(fun a -> (Arg.lhs a, Arg.rhs a, Arg.intent a))) in
+        Some({tid=sub_tid; address=addr; name=sub_name; cconv=calling_convention; args=args;})
+      end
+    | false -> None)))
+
+
+let build_and_return_extern_symbols (project : Project.t) (program : program term) : extern_symbol list =
+  if List.is_empty !extern_symbols then
+    begin
+      build_extern_symbols project program;
+      !extern_symbols
+    end
+  else !extern_symbols
```

Comment:
In the case where there are no extern symbols, the symbol list would be built over and over again.